File: megarover_change_speed_angular/scripts/megarover_change_speed_angular.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Released under the BSD License.import sys

import rospy
import sys
from geometry_msgs.msg import Twist, Vector3

class Megarover_Change_Speed_Angular():
    def __init__(self):
        # Create twist publisher:
        rospy.init_node('megarover_change_speed_angular')
        self._pub_cmd = rospy.Publisher('rover_twist', Twist, queue_size=100)

        # Initialize twist components to zero:
        self._linear_x  = 0.0
        self._angular_z = 0.0

    def get_speed(self):
        print("speed = %f")%self._linear_x
        return self._linear_x
    
    def set_speed(self, speed_p):
        self._linear_x  = speed_p
        print("set speed = %f")%self._linear_x
    
    def get_angular(self):
        print("angular = %f")%self._angular_z
        return self._angular_z
    
    def set_angular(self, angular_p):
        self._angular_z = angular_p
        print("set angular = %f")%self._angular_z
    
    def _send_twist(self):
        linear  = Vector3(self._linear_x, 0.0, 0.0)
        angular = Vector3(0.0, 0.0, self._angular_z)
        r = rospy.Rate(10)
        twist = Twist(linear, angular)
        for i in range(0, 5):
            self._pub_cmd.publish(twist)
            r.sleep()
        print(twist)
    
    speed = property(get_speed, set_speed)
    angular = property(get_angular, set_angular)

def usage():
    return "%s [speed(m/s)] [angular(rad/s)]"%sys.argv[0]

if __name__ == '__main__':
    try:
        if len(sys.argv) == 3:
            mcsa = Megarover_Change_Speed_Angular()
            mcsa.set_speed(float(sys.argv[1]))
            mcsa.set_angular(float(sys.argv[2]))
            mcsa._send_twist()
        else:
            print(usage())
            sys.exit(1)
    except rospy.ROSInterruptException:
        pass

File: megarover_change_speed_angular/src/megarover_change_speed_angular.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Vector3.h>
#include <cmdline.h>

class Megarover_Change_Speed_Angular
{
    public:
        Megarover_Change_Speed_Angular();

        double _linear_x;
        double _angular_z;

        void send_twist();

    private:
        ros::NodeHandle _nh;
        ros::Publisher _tw_pub;
};

Megarover_Change_Speed_Angular::Megarover_Change_Speed_Angular():
    _linear_x(0.0),
    _angular_z(0.0)
{
    _tw_pub = _nh.advertise<geometry_msgs::Twist>("rover_twist", 1);
}

void Megarover_Change_Speed_Angular::send_twist()
{
    geometry_msgs::Twist twist;

    twist.linear.x = _linear_x;
    twist.linear.y = 0.0;
    twist.linear.z = 0.0;
    twist.angular.x = 0.0;
    twist.angular.y = 0.0;
    twist.angular.z = _angular_z;

    ros::Rate loop_late(10);

    /* Send Message */
    for (int i = 0; i < 10; i++){
        _tw_pub.publish(twist);
        ros::spinOnce();
        loop_late.sleep();
    }
}

int main(int argc, char** argv)
{
    /* コマンドラインパーサー */
    cmdline::parser cmdparser;
    cmdparser.add<double>("speed", 's', "Linear speed [m/s]", false, 0.0);
    cmdparser.add<double>("angular", 'a', "Angular velocity [rad/s]", false, 0.0);
    cmdparser.add("help", 'h', "Print Help");

    if (!cmdparser.parse(argc, argv) || cmdparser.exist("help")){
        std::cout << cmdparser.error_full() << cmdparser.usage();
        return EXIT_FAILURE;
    }

    /* ROSノード初期化 */
    ros::init(argc, argv, "megarover_change_speed_angular");
    Megarover_Change_Speed_Angular megarover_change_speed_angular;

    megarover_change_speed_angular._linear_x = cmdparser.get<double>("speed");
    megarover_change_speed_angular._angular_z = cmdparser.get<double>("angular");

    megarover_change_speed_angular.send_twist();
    
    return EXIT_SUCCESS;
}